From 8785559ae5949d5938facae56298d8d153729a1f Mon Sep 17 00:00:00 2001
From: someone5678 <someone5678@users.noreply.github.com>
Date: Tue, 21 May 2024 09:55:51 -0400
Subject: [PATCH 2/4] base: Allow to hide screen capture status from apps [1/2]

* Allow to hide screen capture detection from privacy invasive apps
* This feature will be extended to screenrecord
  in the future release of Android (A15)

Change-Id: Ie2a8e49614dcd8fce2311a644d615ab60670c094
Signed-off-by: someone5678 <someone5678@users.noreply.github.com>
---
 core/java/android/provider/Settings.java          |  6 ++++++
 .../android/server/wm/WindowManagerService.java   | 15 ++++++++++++++-
 2 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 6ad9b3744068..addedf1837d1 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -20650,6 +20650,12 @@ public final class Settings {
         @Readable
         public static final String WINDOW_IGNORE_SECURE = "window_ignore_secure";
 
+        /**
+         * Control whether to hide screen capture status from apps.
+         * @hide
+         */
+        public static final String HIDE_SCREEN_CAPTURE_STATUS = "hide_screen_capture_status";
+        
         /**
          * Setting indicating whether Low Power Standby is enabled, if supported.
          *
diff --git a/services/core/java/com/android/server/wm/WindowManagerService.java b/services/core/java/com/android/server/wm/WindowManagerService.java
index dd4c992494b0..9262a43752b9 100644
--- a/services/core/java/com/android/server/wm/WindowManagerService.java
+++ b/services/core/java/com/android/server/wm/WindowManagerService.java
@@ -10732,7 +10732,6 @@ public class WindowManagerService extends IWindowManager.Stub
         mSurfaceSyncGroupController.markSyncGroupReady(syncGroupToken);
     }
 
-
     /**
      * Must be called when a screenshot is taken via hardware chord.
      *
@@ -10746,7 +10745,9 @@ public class WindowManagerService extends IWindowManager.Stub
                 "notifyScreenshotListeners()")) {
             throw new SecurityException("Requires STATUS_BAR_SERVICE permission");
         }
+
         synchronized (mGlobalLock) {
+            if (shouldHideScreenCapture()) return new ArrayList<>();
             final DisplayContent displayContent = mRoot.getDisplayContent(displayId);
             if (displayContent == null) {
                 return new ArrayList<>();
@@ -10848,6 +10849,10 @@ public class WindowManagerService extends IWindowManager.Stub
     @EnforcePermission(android.Manifest.permission.DETECT_SCREEN_RECORDING)
     @Override
     public boolean registerScreenRecordingCallback(IScreenRecordingCallback callback) {
+        if (shouldHideScreenCapture()) {
+            return false;
+        }
+
         registerScreenRecordingCallback_enforcePermission();
         return mScreenRecordingCallbackController.register(callback);
     }
@@ -10860,6 +10865,10 @@ public class WindowManagerService extends IWindowManager.Stub
     }
 
     void onProcessActivityVisibilityChanged(int uid, boolean visible) {
+        if (shouldHideScreenCapture()) {
+            return;
+        }
+
         mScreenRecordingCallbackController.onProcessActivityVisibilityChanged(uid, visible);
     }
 
@@ -10988,4 +10997,8 @@ public class WindowManagerService extends IWindowManager.Stub
             }
         }
     }
+    private boolean shouldHideScreenCapture() {
+        return Settings.Global.getInt(mContext.getContentResolver(),
+                Settings.Global.HIDE_SCREEN_CAPTURE_STATUS, 0) != 0;
+    }
 }
-- 
2.34.1

